lib/debugger/debugger.dart

```diff
@@ -0,0 +1,1309 @@
+                          span(c: 'octicon octicon-chevron-down'),
+                          span(text: 'Step in'),
+                        ])
+                        ..small(),
+                      stepOver = new PButton(null)
+                        ..add(<CoreElement>[
+                          span(c: 'octicon octicon-chevron-right'),
+                          span(text: 'Step over'),
+                        ])
+                        ..small(),
+                      stepOut = new PButton(null)
+                        ..add(<CoreElement>[
+                          span(c: 'octicon octicon-chevron-up'),
+                          span(text: 'Step out'),
+                        ])
+                        ..small(),
+                    ]),
+                  div()..flex(),
+                  breakOnExceptionControl,
+                ]),
+              sourceArea = div(c: 'section table-border')
+                ..flex()
+                ..layoutVertical()
+                ..add(<CoreElement>[
+                  _sourcePathDiv = div(c: 'source-head'),
+                ]),
+              div(c: 'section table-border secondary-area')
+                ..layoutVertical()
+                ..add(consoleArea.element),
+            ]),
+        ]),
+    ]);
+
+    _sourcePathDiv.setInnerHtml('&nbsp;');
+
+    debuggerState.onSupportsStepping.listen((bool value) {
+      stepOver.enabled = value;
+      stepIn.enabled = value;
+      stepOut.enabled = value;
+    });
+
+    stepOver.click(() => debuggerState.stepOver());
+    stepIn.click(() => debuggerState.stepIn());
+    stepOut.click(() => debuggerState.stepOut());
+
+    final Map<String, dynamic> options = <String, dynamic>{
+      'mode': 'dart',
+      'lineNumbers': true,
+      'gutters': <String>['breakpoints'],
+    };
+    final CodeMirror codeMirror =
+        new CodeMirror.fromElement(sourceArea.element, options: options);
+    codeMirror.setReadOnly(true);
+    // ignore: always_specify_types
+    final codeMirrorElement = _sourcePathDiv.element.parent.children[1];
+    codeMirrorElement.setAttribute('flex', '');
+
+    sourceEditor = new SourceEditor(codeMirror, debuggerState);
+
+    debuggerState.onBreakpointsChanged
+        .listen((List<Breakpoint> breakpoints) async {
+      sourceEditor.setBreakpoints(breakpoints);
+    });
+
+    debuggerState.onPausedChanged.listen((bool paused) async {
+      if (paused) {
+        // Check for async causal frames; fall back to using regular sync frames.
+        final Stack stack = await debuggerState.getStack();
+        List<Frame> frames = stack.asyncCausalFrames ?? stack.frames;
+
+        // Handle breaking-on-exceptions.
+        final InstanceRef reportedException = debuggerState.reportedException;
+        if (reportedException != null && frames.isNotEmpty) {
+          final Frame frame = frames.first;
+
+          final Frame newFrame = new Frame();
+          newFrame.type = frame.type;
```

Comment:
nit: cascades are nice for this case. also consider adding clone constructor to Frame.